module/VuFind/src/VuFind/Controller/Plugin/AbstractRequestBase.php

```diff
@@ -303,6 +316,76 @@ public function getDefaultRequiredDate($checkHolds, $catalog = null,
         return $this->getDateFromArray($dateArray);
     }
 
+    /**
+     * Check if the user-provided dates are valid.
+     *
+     * Returns validated dates and/or an array of validation errors if there are
+     * problems.
+     *
+     * @param string $startDate       User-specified start date
+     * @param string $requiredBy      User-specified required-by date
+     * @param array  $extraHoldFields Hold form fields enabled by
+     * configuration/driver
+     *
+     * @return array
+     */
+    public function validateDates($startDate, $requiredBy, $extraHoldFields)
+    {
+        $result = [
+            'startDateTS' => null,
+            'requiredByTS' => null,
+            'errors' => [],
+        ];
+        if (!in_array('startDate', $extraHoldFields)
+            && !in_array('requiredByDate', $extraHoldFields)
+        ) {
+            return $result;
+        }
+
+        $errors = [];
+        if (in_array('startDate', $extraHoldFields)) {
+            try {
+                $result['startDateTS'] = $startDate
+                    ? $this->dateConverter->convertFromDisplayDate('U', $startDate)
+                    : 0;
+                if ($result['startDateTS'] < time()) {
+                    $errors[] = 'hold_start_date_invalid';
+                }
+            } catch (DateException $e) {
+                $errors[] = 'hold_start_date_invalid';
+            }
+        }
+
+        if (in_array('requiredByDate', $extraHoldFields)) {
+            try {
+                if ($requiredBy) {
+                    $result['requiredByTS'] = $this->dateConverter
+                        ->convertFromDisplayDate('U', $requiredBy, true)
+                        ->setTime(23, 59, 59)
+                        ->getTimestamp();
+                } else {
+                    $result['requiredByTS'] = 0;
+                }
+                if ($result['requiredByTS'] < time()) {
+                    $errors[] = 'hold_required_by_date_invalid';
+                }
+            } catch (DateException $e) {
+                $errors[] = 'hold_required_by_date_invalid';
+            }
+        }
+
+        if (!$errors
+            && in_array('startDate', $extraHoldFields)
+            && in_array('requiredByDate', $extraHoldFields)
+            && $result['startDateTS'] > $result['requiredByTS']
```

Comment:
The plot keeps getting thicker as I read my way down. Maybe an optional validation callback function?

Reply:
...and of course, feel free to tell me this isn't worth the effort. But if that's the case, I think this function belongs in the Holds helper rather than AbstractRequestBase, because as written it seems very holds-specific.

Reply:
This isn't worth the effort. :)

I don't think it's worth it to make the input an array with rules. I think it would just move the logic from the code in the plugin to the array in the caller and would make things more complicated without providing obvious benefits. If we end up needing a more generic validation of a group of dates, we could of course make validateDates a wrapper for a rule-based validation mechanism, but I'd rather leave it to a later point if the need arises. Hold modification will bring the need to validate a "frozen until" date, but I'd still rather add a simple method for that.

I have moved the method to Holds plugin to alleviate the issue of it not being generic. As you rightly pointed out, it's pretty hold-specific anyway. We still have other hold-specific methods (at least the request group -related ones) in AbstractRequestBase, but that's another issue.

Reply:
Thanks, @EreMaijala! It's still on the to-do list to get rid of controller plugins entirely ([VUFIND-1285](https://vufind.org/jira/browse/VUFIND-1285)) since there have been threats of their removal in a future Laminas release... so we can address other refactoring of inappropriately-located methods as part of that project, if and when we get there. (I think this is all really related to the overall goal of making controllers more focused and less "magical").

For now, I'm happy with the relocation of the method. Just one last question before I do the final testing (and hopefully merging) of this work: we've now added the DateConverter as a dependency of all of the request processing helpers, even though Holds is the only one that uses it. Are you planning to apply similar logic to the other types of requests? If so, I'm fine with leaving it where it is so it's ready for future work... but if it's only ever going to be used by Holds, it would probably be cleaner to move it there and give Holds a separate factory.

Reply:
I think it would make sense to add similar checking to other requests as well, so if it's ok to leave it there for now, I'll try to actually make use of it later on.

Reply:
Great, thanks! Not sure if I'll get this tested before next week, but hopefully it won't take too long once I have a free moment.

Reply:
@EreMaijala, I ran the full test suite today, and two integration tests are failing:

<pre>
There were 2 failures:

1) VuFindTest\Mink\IlsActionsTest::testPlaceHold
Selector not found: .modal-body a
Failed asserting that false is true.

/home/dkatz/vufind3/module/VuFind/src/VuFindTest/Integration/MinkTestCase.php:251
/home/dkatz/vufind3/module/VuFind/tests/integration-tests/src/VuFindTest/Mink/IlsActionsTest.php:143
/home/dkatz/vufind3/module/VuFind/tests/integration-tests/src/VuFindTest/Mink/IlsActionsTest.php:260
/home/dkatz/vufind3/module/VuFind/src/VuFindTest/Feature/AutoRetryTrait.php:98

2) VuFindTest\Mink\IlsActionsTest::testHoldsAll
Selector not found: .modal-body a
Failed asserting that false is true.

/home/dkatz/vufind3/module/VuFind/src/VuFindTest/Integration/MinkTestCase.php:251
/home/dkatz/vufind3/module/VuFind/tests/integration-tests/src/VuFindTest/Mink/IlsActionsTest.php:658
/home/dkatz/vufind3/module/VuFind/src/VuFindTest/Feature/AutoRetryTrait.php:98
</pre>

I'm quite short on time today, so I haven't been able to investigate why. If you have a theory, please let me know; otherwise, I'll dig into this further when I have a bit more time. (I'm giving five different presentations at three different events over the next two weeks; once that's done, I'll have a bit more time again!).

Reply:
I suppose placing hold fails, but I'm not sure why. The config change means there'll be the start date field as well, but it should be prepopulated. Other option is that date validation fails, but if the script doesn't change anything, that shouldn't be the case either.

Reply:
I found the issues, will post fixes.

Reply:
The issues should be fixed now. My original intention was to modify VuFindDate to make it support returning a DateTime object, but decided to work around that need for now. Also the start date and required date validation didn't consider today as valid...